base/irrationals.jl

```diff
@@ -60,19 +60,47 @@ AbstractFloat(x::AbstractIrrational) = Float64(x)::Float64
 Float16(x::AbstractIrrational) = Float16(Float32(x)::Float32)
 Complex{T}(x::AbstractIrrational) where {T<:Real} = Complex{T}(T(x))
 
-function _irrational_to_rational(::Type{T}, x::AbstractIrrational) where T<:Integer
-    o = precision(BigFloat)
-    p = 256
-    while true
-        setprecision(BigFloat, p)
-        bx = BigFloat(x)
-        r = rationalize(T, bx, tol=0)
-        if abs(BigFloat(r) - bx) > eps(bx)
-            setprecision(BigFloat, o)
+function _irrational_to_rational_at_current_precision(
+    ::Type{T}, x::AbstractIrrational,
+) where {T <: Integer}
+    bx = BigFloat(x)
+    r = rationalize(T, bx, tol = 0)
+    br = BigFloat(r, precision = precision(BigFloat) + 32)
+    if eps(bx) < abs(br - bx)
+        r
+    else
+        nothing  # Error is too small, repeat with greater precision.
```

Comment:
Don't you mean
```suggestion
        nothing  # Error is too large, repeat with greater precision.
```

Reply:
Actually, no. The idea (both before and after this PR), seems to be to increase `BigFloat` precision until this error becomes large enough. I think that makes sense because a zero error would indicate the `BigFloat` representation of the irrational is exactly represented as a `Rational`, which could indicate the `BigFloat` precision was too small to start with.

The code was originally added in #16527 way back in 2016. @simonbyrne any insight?

Reply:
Actually, now I remember I already came up with what I think is a better algorithm. It's already employed in my package TypeDomainNaturalNumbers.jl, which defines some classes of `AbstractIrrational`s. It relies on the round-up and round-down functionality of `BigFloat`/MPFR to create an interval that encloses an irrational, then compares this to the relevant floating-point interval.

I'll try to apply this to the code here in a follow-up PR, as it's probably better to keep the algorithm as-is in this PR.